### DISCUSSION TITLE
Route Data

### DISCUSSION THREAD

**Discussion by tylerjamesf** [tylerjamesf/MissingNo#7] [General]
Would this be the best way to capture the route information on Pokémon? or would we want relational data to the Pokémon list and how would that look? Assumption that we can use `range(low, high)` and don't need to spell out each level in that area.

```
Route Number, Pokemon, Level Low, Level High, Rate
1, 'Pidgey', 2, 5, 0.5
1, 'Rattata' 2, 4, 0.5
```

Here is what we find on route data from Gen 1 - which I also realized has R, B,Y and that Gen 2 has night and day cycles...

![image](https://user-images.githubusercontent.com/5036869/162483280-af1b6823-9ca7-4110-b531-cb7eaf055712.png)

![image](https://user-images.githubusercontent.com/5036869/162483406-7a7ebb31-30f5-4ca8-a0bb-841e3c6e725a.png)

via [Bulbapedia](https://bulbapedia.bulbagarden.net/wiki/Kanto_Route_1)

**Comment by evokercodex** [comment]
Depends on how crazy you want to go but I think that's a good start. Some things I am thinking about are:

First, how will the application randomly fetch a Pokemon based on it's rate? Is there a finite population and then pick a random number? I am not as familiar with pulling something that's weighted like this. 

Another thing that comes to mind is if we even want higher / lower chances of picking a higher-level Pokémon. As in do I have equal chance of getting a level 2,3,4,5,6? Or higher chance of 2 vs a 6? Might be simpler if discrete records, but that would be a lot of entries if there is a significant range. What would determines the level? Another random function?

Leads me to my next, do want to include spawn location (grass, water, sky) or other "random" qualifiers like time of day (morning, day, night)? Now you're not building an actual game, or are you, that would suggest finding a Rattata in the middle of a lake so it might be worth skipping for now. Time of day might be a fun added challenge and something to visualize / observe, but something would have to randomly pick and then filter from the appropriate weights. Similarly, I also don't think "game" is necessary unless okay with more filters /  random data pickers. 

Depending on complexity & physical file size it might be worthwhile to make separate tables. A suggestion like route_pokemon_xref table and then a spawn table might work as to not duplicate route + pokemon combo.

id  | route | PokemonID
1,1,1
2,1,2
3,2,1
4,3,5

spawn | xrefId, rate
1,1,.5
2,2,.5
3,3,1.0

**Reply by tylerjamesf** [reply]
> First, how will the application randomly fetch a Pokémon based on its rate? Is there a finite population and then pick a random number?

I am not actually sure - but I'm guessing its `random int` - funny enough a lot of the stuff I was reading like that is just RNG ofc based on seeds from game time not computers.

> Another thing that comes to mind is if we even want higher / lower chances of picking a higher-level Pokémon. As in do I have equal chance of getting a level 2,3,4,5,6? Or higher chance of 2 vs a 6? Might be simpler if discrete records, but that would be a lot of entries if there is a significant range. What would determines the level? Another random function?

I honestly think this is the same; if you have range of 2 -6 , it seems you would have an equal chance of finding any in that level range inclusive. Unless somewhere else says otherwise...

> Leads me to my next, do want to include spawn location (grass, water, sky) or other "random" qualifiers like time of day (morning, day, night)?

I think that's a fair question; I was thinking does it make more sense to arrange the data in a way where its more Pokémon is the base and you have PK and FK to subdivide one full list? All and every data point just in case? And for the game - we could just take the difference between the two percentage wise or combine all to get average percentage...

I'm also not familiar with XREF being used for data? How would that work?

**Reply by evokercodex** [reply]
> I am not actually sure - but I'm guessing its `random int` - funny enough a lot of the stuff I was reading like that is just RNG ofc based on seeds from game time not computers.

Seed values are necessary for reproducibility. I'll be honest that's getting out of my wheelhouse.... I thought I saw Python has a get random choice. This might be helpful. https://www.geeksforgeeks.org/how-to-get-weighted-random-choice-in-python/

> I honestly think this is the same; if you have range of 2 -6, it seems you would have an equal chance of finding any in that level range inclusive. Unless somewhere else says otherwise...

I'd agree (even chance/distribution) - Only called it out since their sample indicates it could happen. Like at night you have a chance to get a level 6 vs in the morning / day only a level 2.

**Reply by tylerjamesf** [reply]
Just looked it up a bit - encountering a level of Pokémon is not weighted and neither is the random numbers for the others. Just odds of it out of total sum.

That's a good point though, like would it be just another table for night versus day? or would we create a new entry line so each line of data is unique?
